src/circuit/gadget/sinsemilla/merkle/chip.rs

```diff
@@ -0,0 +1,510 @@
+            CondSwapChip::configure(meta, advices, sinsemilla_config.perm.clone());
+        let lookup_config = LookupRangeCheckConfig::configure(
+            meta,
+            advices[0],
+            sinsemilla_config.constants,
+            sinsemilla_config.generator_table.table_idx,
+            sinsemilla_config.perm.clone(),
+        );
+
+        // This fixed column serves two purposes:
+        //  - Fixing the value of l* for rows in which a Merkle path layer
+        //    is decomposed.
+        //  - Disabling the entire decomposition gate (when set to zero)
+        //    (i.e. replacing a Selector).
+
+        let l_plus_1 = meta.fixed_column();
+
+        // Check that pieces have been decomposed correctly for Sinsemilla hash.
+        // <https://zips.z.cash/protocol/nu5.pdf#orchardmerklecrh>
+        //
+        // a = a_0||a_1 = l_star || (bits 0..=239 of left)
+        // b = b_0||b_1||b_2
+        //   = (bits 240..=249 of left) || (bits 250..=254 of left) || (bits 0..=4 of right)
+        // c = bits 5..=254 of right
+        //
+        // The message pieces `a`, `b`, `c` are constrained by Sinsemilla to be
+        // 250 bits, 20 bits, and 250 bits respectively.
+        //
+        meta.create_gate("Decomposition check", |meta| {
+            let l_plus_1_whole = meta.query_fixed(l_plus_1, Rotation::cur());
+
+            let two_pow_5 = pallas::Base::from_u64(1 << 5);
+            let two_pow_10 = two_pow_5.square();
+
+            // a_whole is constrained by Sinsemilla to be 250 bits.
+            let a_whole = meta.query_advice(advices[0], Rotation::cur());
+            // b_whole is constrained by Sinsemilla to be 20 bits.
+            let b_whole = meta.query_advice(advices[1], Rotation::cur());
+            // c_whole is constrained by Sinsemilla to be 250 bits.
+            let c_whole = meta.query_advice(advices[2], Rotation::cur());
+            let left_node = meta.query_advice(advices[3], Rotation::cur());
+            let right_node = meta.query_advice(advices[4], Rotation::cur());
+
+            // a = a_0||a_1 = l_star || (bits 0..=239 of left)
+            // Check that a_0 = l_star
+            //
+            // z_1 of SinsemillaHash(a) = a_1
+            let z1_a = meta.query_advice(advices[0], Rotation::next());
+            let a_1 = z1_a;
+            // a_0 = a - (a_1 * 2^10)
+            let a_0 = a_whole - a_1.clone() * pallas::Base::from_u64(1 << 10);
+            let l_star_check =
+                a_0 - (l_plus_1_whole.clone() - Expression::Constant(pallas::Base::one()));
+
+            // b = b_0||b_1||b_2
+            //   = (bits 240..=249 of left) || (bits 250..=254 of left) || (bits 0..=4 of right)
+            //
+            //    z_1 of SinsemillaHash(b) = b_1 + 2^5 b_2
+            // => b_0 = b - (z1_b * 2^10)
+            let z1_b = meta.query_advice(advices[1], Rotation::next());
+            // b_1 has been constrained to be 5 bits outside this gate.
```

Comment:
I don't see where this is done (I may have just missed it). Similarly for `b_2`.

Reply:
This is done when `witness_short_check()` is called in `hash_layer()`.